### PR TITLE
 kubelet gets resources from API server of pod origin

### PR DESCRIPTION
**Pull request by h-w-chen**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR is to modify kubelet so that each pod is able to get hold of its essential resources (configmap, secret, service, serviceaccount token) from the API server where the pod spec is kept. This feature is required for daemonset support in scale-out cluster, as daemonsets are defined in system tenant, which could be from different TP servers. 

The main change is around the kube client manager. In current implementation, it keeps track of APi server based on resource's tenant value. In this PR, necessary changes are made to keep track at granularity of pod.

Besides the API server origin tracking, this PR also includes following bug fixes:
* pod env var provisioning takes relevant tenant only when checking for services;
* serviceaccount token manager gets resource with tenant value;

**How it was verified**
at AWS cloud, create 4 EC2 instances to form 2x2 scale-out cluster properly (ensure SCALE_OUT_TP_ENABLE_DAEMONSET is set true at both TPs); after the cluster is stabilized,
1. at each TP server, create daemonset by
``` 
./cluster/kubectl.sh apply -f https://k8s.io/examples/controllers/daemonset.yaml 
```
2. look at the pods created at each TP 
```
./cluster/kubectl.sh get pod -AT -o wide | grep fluentd 
```
3. verify the result indicates that each TP has 2 fluent-* pods running, on both RP, like below
```
system   kube-system   fluentd-elasticsearch-7swls    1/1     Running           0          2m15s   10.246.0.6     ip-172-31-7-34 
system   kube-system   fluentd-elasticsearch-hv687   1/1     Running            0          2m15s   10.245.0.35    ip-172-31-10-97 
```

**Special notes for your reviewer**:
this PR includes a few multi-tenancy related bug fixes, besides the intended feature of kubelet tracking the origin API server of pod-essential resources on basis of pod.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/kubelet_pods.go [hunk 1]
```diff
@@ -527,7 +528,11 @@ func (kl *Kubelet) getServiceEnvVarMap(tenant string, ns string, enableServiceLi
 		// from the master service namespace, even if enableServiceLinks is false.
 		// We also add environment variables for other services in the same
 		// namespace, if enableServiceLinks is true.
-		if service.Namespace == kl.masterServiceNamespace && masterServices.Has(serviceName) {
+		targetedTenant := tenant
+		if !utilfeature.DefaultFeatureGate.Enabled(features.MandatoryArktosNetwork) {
```
Comment (Sindica) [inline pkg/kubelet/kubelet_pods.go:532]: Why is this feature related?

File: pkg/kubelet/secret/secret_manager.go [hunk 2]
```diff
@@ -122,8 +123,8 @@ const (
 //   not there, invalidated or too old, we fetch it from apiserver and refresh the
 //   value in cache; otherwise it is just fetched from cache
 func NewCachingSecretManager(kubeClients []clientset.Interface, getTTL manager.GetObjectTTLFunc) Manager {
-	getSecret := func(tenant, namespace, name string, opts metav1.GetOptions) (runtime.Object, error) {
-		tenantPartitionClient := kubeclientmanager.ClientManager.GetTPClient(kubeClients, tenant)
+	getSecret := func(tenant, namespace, name string, originID int, opts metav1.GetOptions) (runtime.Object, error) {
```
Comment (Sindica) [inline pkg/kubelet/secret/secret_manager.go:126]: Prefer keep same name for this originID. Consider using tenantPartitionID

File: pkg/kubelet/token/token_manager.go [hunk 3]
```diff
@@ -38,13 +40,14 @@ const (
 )
 
 // NewManager returns a new token manager.
-func NewManager(c clientset.Interface) *Manager {
+func NewManager(clients []clientset.Interface) *Manager {
```
Comment (Sindica) [inline pkg/kubelet/token/token_manager.go:43]: Considering using map tpId to clientset. It will be much easier for later dynamic TP/RP support

File: pkg/kubelet/util/manager/manager.go [hunk 4]
```diff
@@ -20,14 +20,15 @@ package manager
...
 )
 
 // Manager is the interface for registering and unregistering
 // objects referenced by pods in the underlying cache and
 // extracting those from that cache if needed.
 type Manager interface {
-	// Get object by its tenant, namespace and name.
-	GetObject(tenant, namespace, name string) (runtime.Object, error)
+	// Get object by its tenant, namespace and name, plus its owner pod.
+	GetObject(tenant, namespace, name string, ownerPod types.UID) (runtime.Object, error)
```
Comment (Sindica) [inline pkg/kubelet/util/manager/manager.go:31]: ownerPodUID

File: pkg/kubelet/util/manager/manager.go [hunk 5]
```diff
@@ -46,16 +47,18 @@ type Manager interface {
...
 // Store is the interface for a object cache that
 // can be used by cacheBasedManager.
+// Since tenant-namespae-name trinity can't uniquely identify a resource in scale-out env,
+// one more factor, origin ID, is needed, representing the origin API server identifier.
 type Store interface {
 	// AddReference adds a reference to the object to the store.
 	// Note that multiple additions to the store has to be allowed
 	// in the implementations and effectively treated as refcounted.
-	AddReference(tenant, namespace, name string)
+	AddReference(tenant, namespace, name string, originId int)
```
Comment (Sindica) [inline pkg/kubelet/util/manager/manager.go:56]: originId is a very generic name. Consider use straight forward name tenantPartitionId

File: pkg/kubelet/volume_host.go [hunk 6]
```diff
@@ -250,15 +250,15 @@ func (kvh *kubeletVolumeHost) GetNodeAllocatable() (v1.ResourceList, error) {
 	return node.Status.Allocatable, nil
 }
 
-func (kvh *kubeletVolumeHost) GetSecretFunc() func(tenant, namespace, name string) (*v1.Secret, error) {
+func (kvh *kubeletVolumeHost) GetSecretFunc() func(tenant, namespace, name string, ownerPod types.UID) (*v1.Secret, error) {
```
Comment (Sindica) [inline pkg/kubelet/volume_host.go:253]: ownerPodUID

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go [hunk 7]
```diff
@@ -474,7 +474,7 @@ func (dswp *desiredStateOfWorldPopulator) deleteProcessedPod(
 // createVolumeSpec creates and returns a mutatable volume.Spec object for the
 // specified volume. It dereference any PVC to get PV objects, if needed.
 // Returns an error if unable to obtain the volume at this time.
-func (dswp *desiredStateOfWorldPopulator) createVolumeSpec(podVolume v1.Volume, podName string, podNamespace string, podTenant string, mountsMap map[string]bool, devicesMap map[string]bool) (*v1.PersistentVolumeClaim, *volume.Spec, string, error) {
+func (dswp *desiredStateOfWorldPopulator) createVolumeSpec(podVolume v1.Volume, podName string, podNamespace string, podTenant string, boundPod types.UID, mountsMap map[string]bool, devicesMap map[string]bool) (*v1.PersistentVolumeClaim, *volume.Spec, string, error) {
```
Comment (Sindica) [inline pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:477]: boundPodUID

File: pkg/volume/configmap/configmap.go [hunk 8]
```diff
@@ -43,7 +43,7 @@ const (
 // configMapPlugin implements the VolumePlugin interface.
 type configMapPlugin struct {
 	host         volume.VolumeHost
-	getConfigMap func(tenant, namespace, name string) (*v1.ConfigMap, error)
+	getConfigMap func(tenant, namespace, name string, ownerPod types.UID) (*v1.ConfigMap, error)
```
Comment (Sindica) [inline pkg/volume/configmap/configmap.go:46]: same here

File: pkg/volume/plugins.go [hunk 9]
```diff
@@ -417,12 +417,12 @@ type VolumeHost interface {
 	GetNodeAllocatable() (v1.ResourceList, error)
 
 	// Returns a function that returns a secret.
-	GetSecretFunc() func(tenant, namespace, name string) (*v1.Secret, error)
+	GetSecretFunc() func(tenant, namespace, name string, ownerPod types.UID) (*v1.Secret, error)
```
Comment (Sindica) [inline pkg/volume/plugins.go:420]: same here

File: pkg/volume/projected/projected.go [hunk 10]
```diff
@@ -48,9 +48,9 @@ const (
 
 type projectedPlugin struct {
 	host                      volume.VolumeHost
-	getSecret                 func(tenant, namespace, name string) (*v1.Secret, error)
-	getConfigMap              func(tenant, namespace, name string) (*v1.ConfigMap, error)
-	getServiceAccountToken    func(namespace, name string, tr *authenticationv1.TokenRequest) (*authenticationv1.TokenRequest, error)
+	getSecret                 func(tenant, namespace, name string, ownerPod types.UID) (*v1.Secret, error)
```
Comment (Sindica) [inline pkg/volume/projected/projected.go:51]: same here

File: pkg/volume/util/operationexecutor/operation_generator.go [hunk 11]
```diff
@@ -857,14 +857,20 @@ func (og *operationGenerator) resizeFileSystem(volumeToMount VolumeToMount, rsOp
 	return true, nil
 }
 
-func (og *operationGenerator) getKubeClient(tenant string) clientset.Interface {
+func (og *operationGenerator) getKubeClient(boundPod types.UID) clientset.Interface {
```
Comment (Sindica) [inline pkg/volume/util/operationexecutor/operation_generator.go:860]: same here

File: pkg/volume/util/operationexecutor/operation_generator.go [hunk 12]
```diff
@@ -1572,7 +1578,7 @@ func (og *operationGenerator) GenerateExpandVolumeFunc(
 			// k8s doesn't have transactions, we can't guarantee that after updating PV - updating PVC will be
 			// successful, that is why all PVCs for which pvc.Spec.Size > pvc.Status.Size must be reprocessed
 			// until they reflect user requested size in pvc.Status.Size
-			kubeClient := og.getKubeClient(pv.Tenant)
+			kubeClient := og.getDefaultKubeClient()
```
Comment (Sindica) [inline pkg/volume/util/operationexecutor/operation_generator.go:1581]: Why default? It this used in RP only?

File: pkg/kubelet/configmap/configmap_manager.go [hunk 13]
```diff
@@ -62,8 +63,8 @@ func NewSimpleConfigMapManager(kubeClients []clientset.Interface) Manager {
 	return &simpleConfigMapManager{kubeClients: kubeClients}
 }
 
-func (s *simpleConfigMapManager) GetConfigMap(tenant, namespace, name string) (*v1.ConfigMap, error) {
-	tenantPartitionClient := kubeclientmanager.ClientManager.GetTPClient(s.kubeClients, tenant)
+func (s *simpleConfigMapManager) GetConfigMap(tenant, namespace, name string, ownerPod types.UID) (*v1.ConfigMap, error) {
```
[the remark this answers is not in the document]
Comment (h-w-chen) [reply pkg/kubelet/configmap/configmap_manager.go:66]: yes will rename as suggested

File: pkg/kubelet/configmap/configmap_manager.go [hunk 14]
```diff
@@ -81,8 +82,8 @@ type configMapManager struct {
 	manager manager.Manager
 }
 
-func (c *configMapManager) GetConfigMap(tenant, namespace, name string) (*v1.ConfigMap, error) {
-	object, err := c.manager.GetObject(tenant, namespace, name)
+func (c *configMapManager) GetConfigMap(tenant, namespace, name string, ownerPod types.UID) (*v1.ConfigMap, error) {
```
[the remark this answers is not in the document]
Comment (h-w-chen) [reply pkg/kubelet/configmap/configmap_manager.go:85]: will rename as suggested

File: pkg/kubelet/configmap/configmap_manager.go [hunk 15]
```diff
@@ -122,8 +123,8 @@ const (
 //   not there, invalidated or too old, we fetch it from apiserver and refresh the
 //   value in cache; otherwise it is just fetched from cache
 func NewCachingConfigMapManager(kubeClients []clientset.Interface, getTTL manager.GetObjectTTLFunc) Manager {
-	getConfigMap := func(tenant, namespace, name string, opts metav1.GetOptions) (runtime.Object, error) {
-		tenantPartitionClient := kubeclientmanager.ClientManager.GetTPClient(kubeClients, tenant)
+	getConfigMap := func(tenant, namespace, name string, origin int, opts metav1.GetOptions) (runtime.Object, error) {
```
[the remark this answers is not in the document]
Comment (h-w-chen) [reply pkg/kubelet/configmap/configmap_manager.go:126]: will rename to originClientID

File: pkg/kubelet/configmap/configmap_manager.go [hunk 16]
```diff
@@ -135,16 +136,16 @@ func NewCachingConfigMapManager(kubeClients []clientset.Interface, getTTL manage
...
 // necessary for registered pods.
 // It implements the following logic:
-// - whenever a pod is created or updated, we start inidvidual watches for all
+// - whenever a pod is created or updated, we start individual watches for all
 //   referenced objects that aren't referenced from other registered pods
 // - every GetObject() returns a value from local cache propagated via watches
 func NewWatchingConfigMapManager(kubeClients []clientset.Interface) Manager {
-	listConfigMap := func(tenant, namespace string, opts metav1.ListOptions) (runtime.Object, error) {
-		tenantPartitionClient := kubeclientmanager.ClientManager.GetTPClient(kubeClients, tenant)
+	listConfigMap := func(tenant, namespace string, originID int, opts metav1.ListOptions) (runtime.Object, error) {
```
[the remark this answers is not in the document]
Comment (h-w-chen) [reply pkg/kubelet/configmap/configmap_manager.go:143]: will rename as originClientID

File: pkg/kubelet/kubeclientmanager/kubeclient_manager.go [hunk 17]
```diff
@@ -23,17 +23,18 @@ import (
...
 )
 
 /*
-This manager keeps a map between tenant name and its corresponding tenant partition apiserver id.
+This manager keeps a map between pod UID and its corresponding tenant partition apiserver id.
 This map is to be used together with kubeTPClients in the Kubelet struct to obtain the correct kubeclient
 */
 type KubeClientManager struct {
-	tenant2api     map[string]int // tenant name -> tenant partition apiserver id
+	puid2api       map[types.UID]int // pod UID -> tenant partition apiserver id
```
[the remark this answers is not in the document]
Comment (h-w-chen) [reply pkg/kubelet/kubeclientmanager/kubeclient_manager.go:37]: puid2api name seems alright; api is short for api server, which is TP in scale-out env.

[the remark this answers is not in the document]
Comment (Sindica) [reply pkg/kubelet/kubeclientmanager/kubeclient_manager.go:37]: Scale out can also have multiple API server in each TP.

File: pkg/kubelet/kubeclientmanager/kubeclient_manager.go [hunk 18]
```diff
@@ -61,13 +62,17 @@ func (manager *KubeClientManager) RegisterTenantSourceServer(source string, ref
 		return
 	}
 
+	if len(ref.UID) == 0 {
+		klog.Warningf("unable to register tenant source for pod with empty UID: tenant='%s', source='%s'", ref.Tenant, source)
```
[the remark this answers is not in the document]
Comment (h-w-chen) [reply pkg/kubelet/kubeclientmanager/kubeclient_manager.go:66]: this source name seems to stand for config source, may taking values like "api0", "api5", even "api". It is better not to change to tpid, as source actually makes more sense here.

File: pkg/kubelet/kubeclientmanager/kubeclient_manager.go [hunk 19]
```diff
@@ -77,28 +82,41 @@ func (manager *KubeClientManager) RegisterTenantSourceServer(source string, ref
...
+		return
+	}
+
+	manager.tenant2apiLock.Lock()
+	defer manager.tenant2apiLock.Unlock()
+
+	delete(manager.puid2api, ref.UID)
+}
+
+func (manager *KubeClientManager) GetTPClient(kubeClients []clientset.Interface, podUID types.UID) clientset.Interface {
```
[the remark this answers is not in the document]
Comment (h-w-chen) [reply pkg/kubelet/kubeclientmanager/kubeclient_manager.go:103]: the collection of kubeclients won't change. What the delete does is to clean up pod's association of kube client (not the client itslef) on when the pod is removed.

File: pkg/kubelet/kubelet_pods.go (hunk 1 above)
Comment (h-w-chen) [reply pkg/kubelet/kubelet_pods.go:532]: good point. should remove this check.

File: pkg/kubelet/kubeclientmanager/kubeclient_manager.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (h-w-chen) [reply pkg/kubelet/kubeclientmanager/kubeclient_manager.go:66]: will add ns+name in log message

File: pkg/kubelet/secret/secret_manager.go (hunk 2 above)
Comment (h-w-chen) [reply pkg/kubelet/secret/secret_manager.go:126]: will use the originClientID as other places.

File: pkg/kubelet/token/token_manager.go (hunk 3 above)
Comment (h-w-chen) [reply pkg/kubelet/token/token_manager.go:43]: as discussed offline, this change is to fix the defect that token_manager unable to support multiple TPs in scale-out env.

File: pkg/kubelet/util/manager/manager.go (hunk 4 above)
Comment (h-w-chen) [reply pkg/kubelet/util/manager/manager.go:31]: yes will rename in the consistent way

File: pkg/kubelet/util/manager/manager.go (hunk 5 above)
Comment (h-w-chen) [reply pkg/kubelet/util/manager/manager.go:56]: yes will rename in the consistent way

File: pkg/kubelet/volume_host.go (hunk 6 above)
Comment (h-w-chen) [reply pkg/kubelet/volume_host.go:253]: yes will rename in the consistent way

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go (hunk 7 above)
Comment (h-w-chen) [reply pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:477]: will rename as suggested

File: pkg/volume/configmap/configmap.go (hunk 8 above)
Comment (h-w-chen) [reply pkg/volume/configmap/configmap.go:46]: will rename in the consistent way

File: pkg/volume/plugins.go (hunk 9 above)
Comment (h-w-chen) [reply pkg/volume/plugins.go:420]: will rename in the consistent way

File: pkg/volume/projected/projected.go (hunk 10 above)
Comment (h-w-chen) [reply pkg/volume/projected/projected.go:51]: will rename in the consistent way

File: pkg/volume/util/operationexecutor/operation_generator.go (hunk 11 above)
Comment (h-w-chen) [reply pkg/volume/util/operationexecutor/operation_generator.go:860]: will rename in the consistent way

File: pkg/volume/util/operationexecutor/operation_generator.go (hunk 12 above)
Comment (h-w-chen) [reply pkg/volume/util/operationexecutor/operation_generator.go:1581]: this will assumed to be used in TP (as part of KCM) only. the default kube client is pointing to TP. see line 868-869 comment.

However, if this assumption is nt valid, this par tof code need to change.

File: pkg/kubelet/kubeclientmanager/kubeclient_manager.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (h-w-chen) [reply pkg/kubelet/kubeclientmanager/kubeclient_manager.go:37]: OK got it. in that case, what we care here is actually the api server; as long as we get hold of the proper api server (by proper client), it is alright.  will change the comment accordingly.

File: pkg/kubelet/kubeclientmanager/kubeclient_manager.go [hunk 20]
```diff
@@ -23,17 +23,18 @@ import (
...
 
 /*
-This manager keeps a map between tenant name and its corresponding tenant partition apiserver id.
+This manager keeps a map between pod UID and its corresponding tenant partition apiserver id.
 This map is to be used together with kubeTPClients in the Kubelet struct to obtain the correct kubeclient
 */
 type KubeClientManager struct {
-	tenant2api     map[string]int // tenant name -> tenant partition apiserver id
+	puid2api       map[types.UID]int // pod UID -> apiserver id
 	tenant2apiLock sync.RWMutex
```
Comment (Sindica) [inline pkg/kubelet/kubeclientmanager/kubeclient_manager.go:38]: The lock should be renamed

Comment (h-w-chen) [reply pkg/kubelet/kubeclientmanager/kubeclient_manager.go:38]: will change as suggested

File: pkg/kubelet/kubeclientmanager/kubeclient_manager.go [hunk 21]
```diff
@@ -77,28 +82,41 @@ func (manager *KubeClientManager) RegisterTenantSourceServer(source string, ref
 		return
 	}
 
-	if _, ok := manager.tenant2api[key]; !ok {
-		manager.tenant2api[key] = clientId
-		klog.Infof("added %d to the map, map has %+v", clientId, manager.tenant2api)
+	if _, ok := manager.puid2api[ref.UID]; !ok {
```
Comment (Sindica) [inline pkg/kubelet/kubeclientmanager/kubeclient_manager.go:85]: Question: should we compare value if podUID is found in the map?

File: pkg/kubelet/kubeclientmanager/kubeclient_manager.go [hunk 22]
```diff
@@ -77,28 +82,41 @@ func (manager *KubeClientManager) RegisterTenantSourceServer(source string, ref
...
+		manager.puid2api[ref.UID] = clientId
+		klog.V(6).Infof("added %d to the map, map has %+v", clientId, manager.puid2api)
 	}
 }
 
-func (manager *KubeClientManager) GetTPClient(kubeClients []clientset.Interface, tenant string) clientset.Interface {
+func (manager *KubeClientManager) UnregisterTenantSourceServer(ref *v1.Pod) {
```
Comment (Sindica) [inline pkg/kubelet/kubeclientmanager/kubeclient_manager.go:91]: Rename to UnregisterPodSourceServer?

File: pkg/kubelet/kubeclientmanager/kubeclient_manager.go [hunk 23]
```diff
@@ -77,28 +82,41 @@ func (manager *KubeClientManager) RegisterTenantSourceServer(source string, ref
...
+	if len(ref.UID) == 0 {
+		klog.Warningf("unable to unregister tenant source for pod with empty UID: tenant='%s'", ref.Tenant)
```
Comment (Sindica) [inline pkg/kubelet/kubeclientmanager/kubeclient_manager.go:93]: unable to unregister api source for pod?

File: pkg/kubelet/network/dns/dns.go [hunk 24]
```diff
@@ -116,7 +116,8 @@ func (c *Configurer) getClusterDNS(pod *v1.Pod) ([]net.IP, error) {
 	if len(c.serviceListers) > 1 {
 		klog.V(4).Infof("Need locate which serviceLister is correct one in multi-servceListers")
 		// To get correct index of serviceLister based on Pod's tenant information
-		index = c.clientManager.PickClient(pod.Tenant)
+		// todo: remove param of tenant after pod specific origin tracking impl
```
Comment (Sindica) [inline pkg/kubelet/network/dns/dns.go:119]: Can this todo be done now? Or just leave tenant for accuracy?

File: pkg/kubelet/token/token_manager.go [hunk 25]
```diff
@@ -38,13 +40,14 @@ const (
...
 	m := &Manager{
-		getToken: func(name, namespace string, tr *authenticationv1.TokenRequest) (*authenticationv1.TokenRequest, error) {
-			if c == nil {
+		getToken: func(tenant, name, namespace string, ownerPodUID types.UID, tr *authenticationv1.TokenRequest) (*authenticationv1.TokenRequest, error) {
```
Comment (Sindica) [inline pkg/kubelet/token/token_manager.go:45]: Why add tenant as well as podUID?

File: pkg/kubelet/util/manager/cache_based_manager.go [hunk 26]
```diff
@@ -24,26 +24,28 @@ import (
...
 	"k8s.io/apimachinery/pkg/util/sets"
+	storageetcd3 "k8s.io/apiserver/pkg/storage/etcd3"
+	"k8s.io/kubernetes/pkg/kubelet/kubeclientmanager"
+	"k8s.io/kubernetes/pkg/kubelet/util"
 )
 
 // GetObjectTTLFunc defines a function to get value of TTL.
 type GetObjectTTLFunc func() (time.Duration, bool)
 
 // GetObjectFunc defines a function to get object with a given tenant, namespace and name.
```
Comment (Sindica) [inline pkg/kubelet/util/manager/cache_based_manager.go:41]: New argument added, need to give name or update comment

File: pkg/kubelet/util/manager/watch_based_manager.go [hunk 27]
```diff
@@ -42,8 +42,8 @@ import (
 	"k8s.io/apimachinery/pkg/watch"
 )
 
-type listObjectFunc func(string, string, metav1.ListOptions) (runtime.Object, error)
-type watchObjectFunc func(string, string, metav1.ListOptions) (watch.Interface, error)
+type listObjectFunc func(string, string, int, metav1.ListOptions) (runtime.Object, error)
```
Comment (Sindica) [inline pkg/kubelet/util/manager/watch_based_manager.go:45]: Can we add comment or name for arguments?

File: pkg/volume/util/operationexecutor/operation_generator.go [hunk 28]
```diff
@@ -857,14 +857,20 @@ func (og *operationGenerator) resizeFileSystem(volumeToMount VolumeToMount, rsOp
...
 	} else {
-		return kubeclientmanager.ClientManager.GetTPClient(og.kubeTenantPartitionClients, tenant)
+		return kubeclientmanager.ClientManager.GetTPClient(og.kubeTenantPartitionClients, boundPodUID)
 	}
 }
 
+// supposed to be called by controller (e.g. expandController) inside of KCM,
+// assuming the controller is within TP env in scale-out cluster.
+func (og *operationGenerator) getDefaultKubeClient() clientset.Interface {
+	return og.kubeTenantPartitionClients[0]
```
Comment (Sindica) [inline pkg/volume/util/operationexecutor/operation_generator.go:871]: Is operationGenerator used both in TP and RP? If not, it should not have multiple TPClients in TP controllers.

File: pkg/kubelet/kubeclientmanager/kubeclient_manager.go [hunk 29]
```diff
@@ -49,9 +50,9 @@ func NewKubeClientManager() {
 func newKubeClientManagerFunc() func() {
 	return func() {
 		ClientManager = &KubeClientManager{
-			tenant2api: make(map[string]int),
+			puid2api: make(map[types.UID]int),
 		}
-		klog.Infof("kubeclient manager initialized %v", ClientManager)
+		klog.V(2).Infof("kubeclient manager initialized %v", ClientManager)
 	}
 }
```
[the remark this answers is not in the document]
Comment (h-w-chen) [reply pkg/kubelet/kubeclientmanager/kubeclient_manager.go:58]: 1. will change method name as suggested.
2. yes it is better to be, as it may indicate the or condition branch.

File: pkg/kubelet/kubeclientmanager/kubeclient_manager.go [hunk 30]
```diff
@@ -61,13 +62,17 @@ func (manager *KubeClientManager) RegisterTenantSourceServer(source string, ref
 		return
 	}
 
-	manager.tenant2apiLock.Lock()
-	defer manager.tenant2apiLock.Unlock()
+	if len(ref.UID) == 0 {
+		klog.Warningf("unable to register tenant source for pod with empty UID: tenant='%s', namespace='%s', name='%s', source='%s'", ref.Tenant, ref.Namespace, ref.Name, source)
```
[the remark this answers is not in the document]
Comment (h-w-chen) [reply pkg/kubelet/kubeclientmanager/kubeclient_manager.go:66]: will rephrase.

File: pkg/kubelet/kubeclientmanager/kubeclient_manager.go (hunk 21 above)
Comment (h-w-chen) [reply pkg/kubelet/kubeclientmanager/kubeclient_manager.go:85]: will add warning log for this case.

File: pkg/kubelet/kubeclientmanager/kubeclient_manager.go (hunk 22 above)
Comment (h-w-chen) [reply pkg/kubelet/kubeclientmanager/kubeclient_manager.go:91]: will rename as suggested

File: pkg/kubelet/kubeclientmanager/kubeclient_manager.go (hunk 23 above)
Comment (h-w-chen) [reply pkg/kubelet/kubeclientmanager/kubeclient_manager.go:93]: will rephrase

File: pkg/kubelet/network/dns/dns.go (hunk 24 above)
Comment (h-w-chen) [reply pkg/kubelet/network/dns/dns.go:119]: already did. will remove the obsolete comment. thanks for pointing it out.

File: pkg/kubelet/token/token_manager.go (hunk 25 above)
Comment (h-w-chen) [reply pkg/kubelet/token/token_manager.go:45]: token param is needed for multi-tenant call. The func body used to call the single-tenanted call, which was defective.

File: pkg/kubelet/util/manager/cache_based_manager.go (hunk 26 above)
Comment (h-w-chen) [reply pkg/kubelet/util/manager/cache_based_manager.go:41]: will add based on feedback

File: pkg/kubelet/util/manager/watch_based_manager.go (hunk 27 above)
Comment (h-w-chen) [reply pkg/kubelet/util/manager/watch_based_manager.go:45]: will add arg names.

File: pkg/volume/util/operationexecutor/operation_generator.go (hunk 28 above)
Comment (h-w-chen) [reply pkg/volume/util/operationexecutor/operation_generator.go:871]: it is used both in kubelet(which is at RP) and KCM controller (which is supposedly at TP)

**Review by Sindica — APPROVED** [review]
/lgtm
